File: app/src/main/java/org/fdroid/fdroid/installer/SessionInstallManager.java
package org.fdroid.fdroid.installer;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.IntentSender;
import android.content.pm.InstallSourceInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageInstaller;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.core.content.ContextCompat;
import androidx.core.util.ObjectsCompat;
import androidx.documentfile.provider.DocumentFile;

import org.apache.commons.io.IOUtils;
import org.fdroid.fdroid.Preferences;
import org.fdroid.fdroid.Utils;
import org.fdroid.fdroid.data.Apk;
import org.fdroid.fdroid.data.App;
import org.fdroid.fdroid.net.DownloaderService;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.concurrent.TimeUnit;

public class SessionInstallManager extends BroadcastReceiver {

    private static final String TAG = "SessionInstallManager";
    private static final String INSTALLER_ACTION_INSTALL =
            "org.fdroid.fdroid.installer.SessionInstallManager.install";
    private static final String INSTALLER_ACTION_UNINSTALL =
            "org.fdroid.fdroid.installer.SessionInstallManager.uninstall";
    /**
     * An intent extra needed only due to a bug in Android 12 (#2599) where our App parcelable in the confirmation
     * intent causes a crash.
     * To prevent this, we wrap the App and Apk parcelables in this bundle.
     */
    private static final String EXTRA_BUNDLE =
            "org.fdroid.fdroid.installer.SessionInstallManager.bundle";

    private final Context context;

    // Used to cache isStockXiaomi() to prevent repeat PackageManager calls
    @Nullable
    private static Boolean isStockXiaomi = null;

    public SessionInstallManager(Context context) {
        this.context = context;
        ContextCompat.registerReceiver(context, this, new IntentFilter(INSTALLER_ACTION_INSTALL),
                ContextCompat.RECEIVER_NOT_EXPORTED);
        ContextCompat.registerReceiver(context, this, new IntentFilter(INSTALLER_ACTION_UNINSTALL),
                ContextCompat.RECEIVER_NOT_EXPORTED);
        PackageInstaller installer = context.getPackageManager().getPackageInstaller();
        // abandon old sessions, because there's a limit
        // that will throw IllegalStateException when we try to open new sessions
        Utils.runOffUiThread(() -> {
            for (PackageInstaller.SessionInfo session : installer.getMySessions()) {
                Utils.debugLog(TAG, "Abandon session " + session.getSessionId());
                try {
                    installer.abandonSession(session.getSessionId());
                } catch (SecurityException e) {
                    Log.e(TAG, "Error abandoning session: ", e);
                }
            }
        });
    }

    @WorkerThread
    public void install(App app, Apk apk, Uri localApkUri, Uri canonicalUri) {
        DocumentFile documentFile = ObjectsCompat.requireNonNull(DocumentFile.fromSingleUri(context, localApkUri));
        long size = documentFile.length();
        Log.i(TAG, "Installing " + documentFile.getUri() + " with size " + size + " bytes");
        PackageInstaller.SessionParams params = getSessionParams(app, size);
        PackageInstaller installer = context.getPackageManager().getPackageInstaller();
        try {
            int sessionId = installer.createSession(params);
            ContentResolver contentResolver = context.getContentResolver();
            try (PackageInstaller.Session session = installer.openSession(sessionId)) {
                try (InputStream inputStream = contentResolver.openInputStream(localApkUri)) {
                    try (OutputStream outputStream = session.openWrite(app.packageName, 0, size)) {
                        IOUtils.copy(inputStream, outputStream);
                        session.fsync(outputStream);
                    }
                }
                IntentSender sender = getInstallIntentSender(sessionId, app, apk, canonicalUri);
                // wait for install constraints, if they can be used
                if (Build.VERSION.SDK_INT >= 34 && canUseInstallConstraints(app.packageName)) {
                    // we are allowed, so wait for constraints to do gentle update
                    PackageInstaller.InstallConstraints constraints =
                            PackageInstaller.InstallConstraints.GENTLE_UPDATE;
                    long timeout = TimeUnit.HOURS.toMillis(3);
                    Log.i(TAG, "Committing session using install constraints...");
                    installer.commitSessionAfterInstallConstraintsAreMet(sessionId, sender, constraints, timeout);
                } else {
                    Log.i(TAG, "Committing session without install constraints...");
                    session.commit(sender);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "I/O Error during install session: ", e);
            Installer.sendBroadcastInstall(context, canonicalUri, Installer.ACTION_INSTALL_INTERRUPTED, app, apk,
                    null, e.getLocalizedMessage());
        }
    }

    @NonNull
    private static PackageInstaller.SessionParams getSessionParams(App app, long size) {
        PackageInstaller.SessionParams params =
                new PackageInstaller.SessionParams(PackageInstaller.SessionParams.MODE_FULL_INSTALL);
        params.setAppPackageName(app.packageName);
        params.setSize(size);
        params.setInstallLocation(PackageInfo.INSTALL_LOCATION_AUTO);
        if (Build.VERSION.SDK_INT >= 26) {
            params.setInstallReason(PackageManager.INSTALL_REASON_USER);
        }
        if (Build.VERSION.SDK_INT >= 31) {
            params.setRequireUserAction(PackageInstaller.SessionParams.USER_ACTION_NOT_REQUIRED);
        }
        if (Build.VERSION.SDK_INT >= 33) {
            params.setPackageSource(PackageInstaller.PACKAGE_SOURCE_STORE);
        }
        if (Build.VERSION.SDK_INT >= 34) {
            // Once the update ownership enforcement is enabled,
            // the other installers will need the user action to update the package
            // even if the installers have been granted the INSTALL_PACKAGES permission.
            // The update ownership enforcement can only be enabled on initial installation.
            // Set this to true on package update is a no-op.
            params.setRequestUpdateOwnership(true);
        }
        return params;
    }

    private boolean canUseInstallConstraints(String packageName) {
        // We had too many complaints about gentle updates failing to update apps:
        // * with running foreground services
        // * used default keyboards
        // * being media players
        // So we'll disable this feature for now until it works properly.
        if (true) return false;

        String ourPackageName = context.getPackageName();
        if (Build.VERSION.SDK_INT < 34 || packageName.equals(ourPackageName)) return false;
        try {
            InstallSourceInfo sourceInfo = context.getPackageManager().getInstallSourceInfo(packageName);
            return ourPackageName.equals(sourceInfo.getInstallingPackageName()) ||
                    ourPackageName.equals(sourceInfo.getUpdateOwnerPackageName());
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    @WorkerThread
    public void uninstall(String packageName) {
        PackageInstaller installer = context.getPackageManager().getPackageInstaller();
        installer.uninstall(packageName, getUninstallIntentSender(packageName));
    }

    private IntentSender getInstallIntentSender(int sessionId, App app, Apk apk, Uri canonicalUri) {
        Intent broadcastIntent = new Intent(INSTALLER_ACTION_INSTALL);
        broadcastIntent.setPackage(context.getPackageName());
        broadcastIntent.putExtra(PackageInstaller.EXTRA_SESSION_ID, sessionId);
        Bundle bundle = new Bundle();
        bundle.putParcelable(Installer.EXTRA_APP, app);
        bundle.putParcelable(Installer.EXTRA_APK, apk);
        broadcastIntent.putExtra(EXTRA_BUNDLE, bundle);
        broadcastIntent.putExtra(DownloaderService.EXTRA_CANONICAL_URL, canonicalUri);
        // we are stuffing this intent pretty full, hopefully won't run into the size limit
        broadcastIntent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND);
        // intent flag needs to be mutable, otherwise the intent has no extras
        int flags = Build.VERSION.SDK_INT >= 31 ?
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_MUTABLE :
                PendingIntent.FLAG_UPDATE_CURRENT;
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context, sessionId, broadcastIntent, flags);
        return pendingIntent.getIntentSender();
    }

    private IntentSender getUninstallIntentSender(String packageName) {
        Intent broadcastIntent = new Intent(INSTALLER_ACTION_UNINSTALL);
        broadcastIntent.setPackage(context.getPackageName());
        broadcastIntent.putExtra(PackageInstaller.EXTRA_PACKAGE_NAME, packageName);
        broadcastIntent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND);
        // intent flag needs to be mutable, otherwise the intent has no extras
        int flags = Build.VERSION.SDK_INT >= 31 ?
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_MUTABLE :
                PendingIntent.FLAG_UPDATE_CURRENT;
        PendingIntent pendingIntent =
                PendingIntent.getBroadcast(context, packageName.hashCode(), broadcastIntent, flags);
        return pendingIntent.getIntentSender();
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (INSTALLER_ACTION_INSTALL.equals(intent.getAction())) {
            onInstallReceived(intent);
        } else if (INSTALLER_ACTION_UNINSTALL.equals(intent.getAction())) {
            onUninstallReceived(intent);
        } else {
            throw new IllegalStateException("Unsupported broadcast action: " + intent.getAction());
        }
    }

    private void onInstallReceived(Intent intent) {
        int sessionId = intent.getIntExtra(PackageInstaller.EXTRA_SESSION_ID, -1);
        Intent confirmIntent = intent.getParcelableExtra(Intent.EXTRA_INTENT);

        Bundle bundle = intent.getBundleExtra(EXTRA_BUNDLE);
        App app = bundle.getParcelable(Installer.EXTRA_APP);
        Apk apk = bundle.getParcelable(Installer.EXTRA_APK);
        Uri canonicalUri = intent.getParcelableExtra(DownloaderService.EXTRA_CANONICAL_URL);

        int status = intent.getIntExtra(PackageInstaller.EXTRA_STATUS, Integer.MIN_VALUE);
        String msg = intent.getStringExtra(PackageInstaller.EXTRA_STATUS_MESSAGE);

        Log.i(TAG, "Received install broadcast for " + app.packageName + " " + status + ": " + msg);

        if (status == PackageInstaller.STATUS_SUCCESS) {
            String action = Installer.ACTION_INSTALL_COMPLETE;
            Installer.sendBroadcastInstall(context, canonicalUri, action, app, apk, null, null);
        } else if (status == PackageInstaller.STATUS_PENDING_USER_ACTION) {
            int flags = Build.VERSION.SDK_INT >= 31 ?
                    PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE :
                    PendingIntent.FLAG_UPDATE_CURRENT;
            PendingIntent pendingIntent = PendingIntent.getActivity(context, sessionId, confirmIntent, flags);
            String action = Installer.ACTION_INSTALL_USER_INTERACTION;
            Installer.sendBroadcastInstall(context, canonicalUri, action, app, apk, pendingIntent, null);
        } else {
            // show no message when user actively aborted
            String m = status == PackageInstaller.STATUS_FAILURE_ABORTED ? null : msg;
            String action = Installer.ACTION_INSTALL_INTERRUPTED;
            Installer.sendBroadcastInstall(context, canonicalUri, action, app, apk, null, m);
        }
    }

    private void onUninstallReceived(Intent intent) {
        String packageName = intent.getStringExtra(PackageInstaller.EXTRA_PACKAGE_NAME);
        Intent confirmIntent = intent.getParcelableExtra(Intent.EXTRA_INTENT);
        int status = intent.getIntExtra(PackageInstaller.EXTRA_STATUS, Integer.MIN_VALUE);
        String msg = intent.getStringExtra(PackageInstaller.EXTRA_STATUS_MESSAGE);

        Log.i(TAG, "Received uninstall broadcast for " + packageName + " " + status + ": " + msg);

        if (status == PackageInstaller.STATUS_SUCCESS) {
            String action = Installer.ACTION_UNINSTALL_COMPLETE;
            sendBroadcastUninstall(packageName, action, null, null);
        } else if (status == PackageInstaller.STATUS_PENDING_USER_ACTION) {
            int flags = Build.VERSION.SDK_INT >= 31 ?
                    PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE :
                    PendingIntent.FLAG_UPDATE_CURRENT;
            PendingIntent pendingIntent =
                    PendingIntent.getActivity(context, packageName.hashCode(), confirmIntent, flags);
            String action = Installer.ACTION_UNINSTALL_USER_INTERACTION;
            sendBroadcastUninstall(packageName, action, pendingIntent, null);
        } else {
            // show no message when user actively aborted
            String m = status == PackageInstaller.STATUS_FAILURE_ABORTED ? null : msg;
            String action = Installer.ACTION_UNINSTALL_INTERRUPTED;
            sendBroadcastUninstall(packageName, action, null, m);
        }
    }

    private void sendBroadcastUninstall(String packageName, String action, @Nullable PendingIntent pendingIntent,
                                        @Nullable String errorMessage) {
        App app = new App();
        app.packageName = packageName;
        Apk apk = new Apk();
        apk.packageName = packageName;
        Installer.sendBroadcastUninstall(context, app, apk, action, pendingIntent, errorMessage);
    }

    /**
     * Returns true if the {@link SessionInstaller} can be used on this device.
     */
    public static boolean canBeUsed(Context context) {
        // In case of bugs, let the user disable this while it is still beta.
        if (Preferences.get().forceOldInstaller()) return false;
        // We could use the SessionInstaller also on lower versions,
        // but the benefit of unattended updates only starts with SDK 31.
        // Before the extra bugs it has aren't worth it.
        if (Build.VERSION.SDK_INT < 31) return false;
        // Xiaomi MIUI (at least in version 12) is known to break the PackageInstaller API in several ways.
        // Disabling MIUI "optimizations" in developer options fixes it, but we can't ask users to do this (bad UX).
        // Therefore, we have no choice, but to disable it completely for those devices.
        // See: https://github.com/vvb2060/PackageInstallerTest
        if (isStockXiaomi(context)) return false;
        // We don't use SessionInstaller, if PrivilegedInstaller can be used instead.
        // This is the last check, because it is the most expensive one
        // getting PackageInfo and doing service binding.
        return !PrivilegedInstaller.isDefault(context);
    }

    private static boolean isStockXiaomi(Context context) {
        if (isStockXiaomi == null) {
            boolean xiaomiPhone = "Xiaomi".equalsIgnoreCase(Build.BRAND) || "Redmi".equalsIgnoreCase(Build.BRAND);
            if (xiaomiPhone) {
                // Calls for non-installed packages take longer than installed ones
                // MIUI OS will result in one call
                // Non-MIUI OS will result in two calls
                if (Utils.getPackageInfo(context, "com.miui.securitycenter") != null) {
                    isStockXiaomi = true;
                } else {
                    isStockXiaomi = Utils.getPackageInfo(context, "com.miui.packageinstaller") != null;
                }
            } else {
                isStockXiaomi = false;
            }
        }
        return isStockXiaomi;
    }

    /**
     * If this returns true, we can use
     * {@link android.content.pm.PackageInstaller.SessionParams#setRequireUserAction(int)} with false,
     * thus updating the app with the given targetSdk without user action.
     */
    public static boolean isTargetSdkSupported(int targetSdk) {
        if (Build.VERSION.SDK_INT < 31) return false; // not supported below Android 12
        if (Build.VERSION.SDK_INT == 31 && targetSdk >= 29) return true;
        if (Build.VERSION.SDK_INT == 32 && targetSdk >= 29) return true;
        if (Build.VERSION.SDK_INT == 33 && targetSdk >= 30) return true;
        if (Build.VERSION.SDK_INT == 34 && targetSdk >= 31) return true;
        // This needs to be adjusted as new Android versions are released
        // https://developer.android.com/reference/android/content/pm/PackageInstaller.SessionParams#setRequireUserAction(int)
        // https://cs.android.com/android/platform/superproject/+/android-16.0.0_r2:frameworks/base/services/core/java/com/android/server/pm/PackageInstallerSession.java;l=329;drc=73caa0299d9196ddeefe4f659f557fb880f6536d
        // current code requires targetSdk 33 on SDK 35+
        return Build.VERSION.SDK_INT >= 35 && targetSdk >= 33;
    }
}

File: next/src/main/kotlin/org/fdroid/install/SessionInstallManager.kt
package org.fdroid.install

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_IMMUTABLE
import android.app.PendingIntent.FLAG_MUTABLE
import android.app.PendingIntent.FLAG_UPDATE_CURRENT
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.IntentSender
import android.content.pm.PackageInfo
import android.content.pm.PackageInstaller
import android.content.pm.PackageInstaller.EXTRA_PACKAGE_NAME
import android.content.pm.PackageInstaller.EXTRA_SESSION_ID
import android.content.pm.PackageInstaller.SessionParams
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.icu.util.ULocale
import android.os.Build.VERSION.SDK_INT
import androidx.annotation.RequiresApi
import androidx.annotation.WorkerThread
import androidx.core.content.ContextCompat.RECEIVER_NOT_EXPORTED
import androidx.core.content.ContextCompat.registerReceiver
import androidx.core.os.LocaleListCompat
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import mu.KotlinLogging
import org.fdroid.LocaleChooser.getBestLocale
import org.fdroid.database.AppMetadata
import org.fdroid.utils.IoDispatcher
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

@Singleton
class SessionInstallManager @Inject constructor(
    @param:ApplicationContext private val context: Context,
    @param:IoDispatcher private val coroutineScope: CoroutineScope,
) {

    private val log = KotlinLogging.logger { }
    private val installer = context.packageManager.packageInstaller

    companion object {
        private const val ACTION_INSTALL = "org.fdroid.install.SessionInstallManager.install"

        /**
         * If this returns true, we can use
         * [SessionParams.setRequireUserAction] with false,
         * thus updating the app with the given targetSdk without user action.
         */
        fun isAutoUpdateSupported(targetSdk: Int): Boolean {
            if (SDK_INT < 31) return false // not supported below Android 12

            if (SDK_INT == 31 && targetSdk >= 29) return true
            if (SDK_INT == 32 && targetSdk >= 29) return true
            if (SDK_INT == 33 && targetSdk >= 30) return true
            if (SDK_INT == 34 && targetSdk >= 31) return true
            // This needs to be adjusted as new Android versions are released
            // https://developer.android.com/reference/android/content/pm/PackageInstaller.SessionParams#setRequireUserAction(int)
            // https://cs.android.com/android/platform/superproject/+/android-16.0.0_r2:frameworks/base/services/core/java/com/android/server/pm/PackageInstallerSession.java;l=329;drc=73caa0299d9196ddeefe4f659f557fb880f6536d
            // current code requires targetSdk 33 on SDK 35+
            return SDK_INT >= 35 && targetSdk >= 33
        }
    }

    init {
        // abandon old sessions, because there's a limit
        // that will throw IllegalStateException when we try to open new sessions
        coroutineScope.launch {
            for (session in installer.mySessions) {
                log.debug { "Abandon session ${session.sessionId} for ${session.appPackageName}" }
                try {
                    installer.abandonSession(session.sessionId)
                } catch (e: SecurityException) {
                    log.error(e) { "Error abandoning session: " }
                }
            }
        }
    }

    /**
     * Requests installation pre-approval (if available on this device).
     */
    suspend fun requestPreapproval(app: AppMetadata, icon: Bitmap?): PreApprovalResult {
        return if (SDK_INT >= 34) {
            try {
                preapproval(app, icon)
            } catch (e: Exception) {
                log.error(e) { "Error requesting pre-approval: " }
                PreApprovalResult.Error("${e::class.java.simpleName} ${e.message}")
            }
        } else {
            PreApprovalResult.NotSupported
        }
    }

    @RequiresApi(34)
    private suspend fun preapproval(
        app: AppMetadata,
        icon: Bitmap?,
    ): PreApprovalResult = suspendCancellableCoroutine { cont ->
        val params = getSessionParams(app.packageName)
        val sessionId = installer.createSession(params)
        log.info { "Opened session $sessionId" }
        val name = app.name.getBestLocale(LocaleListCompat.getDefault()) ?: ""

        val receiver = InstallBroadcastReceiver(sessionId) { status, intent, msg ->
            when (status) {
                PackageInstaller.STATUS_SUCCESS -> {
                    cont.resume(PreApprovalResult.Success(sessionId))
                    context.unregisterReceiver(this)
                }
                PackageInstaller.STATUS_PENDING_USER_ACTION -> {
                    val flags = FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE
                    val pendingIntent =
                        PendingIntent.getActivity(context, sessionId, intent, flags)
                    // There should be no bugs on Android versions where this is supported
                    // and we should be in the foreground right now,
                    // so fire up intent here and now.
                    pendingIntent.send()
                }
                else -> { // some error, can't help it now, continue
                    if (status == PackageInstaller.STATUS_FAILURE_ABORTED) {
                        cont.resume(PreApprovalResult.UserAborted)
                    } else {
                        cont.resume(PreApprovalResult.Error(msg))
                    }
                    context.unregisterReceiver(this)
                }
            }
        }
        registerReceiver(
            context,
            receiver,
            IntentFilter(ACTION_INSTALL),
            RECEIVER_NOT_EXPORTED
        )
        cont.invokeOnCancellation {
            log.info { "Pre-approval cancelled." }
            context.unregisterReceiver(receiver)
        }

        installer.openSession(sessionId).use { session ->
            log.info { "app name locales: ${app.name} using: ${ULocale.getDefault()}" }
            val details = PackageInstaller.PreapprovalDetails.Builder()
                .setPackageName(app.packageName)
                .setLabel(name)
                .setLocale(ULocale.getDefault()) // TODO get the real one used for label
                .apply { if (icon != null) setIcon(icon) }
                .build()
            val sender = getInstallIntentSender(sessionId, app.packageName)
            session.requestUserPreapproval(details, sender)
        }
        sessionId
    }

    @WorkerThread
    @SuppressLint("RequestInstallPackagesPolicy")
    suspend fun install(
        sessionId: Int?,
        packageName: String,
        state: InstallStateWithInfo,
        apkFile: File,
    ): InstallState = suspendCancellableCoroutine { cont ->
        val size = apkFile.length()
        log.info { "Installing ${apkFile.name} with size $size bytes" }

        val sessionId = try {
            if (sessionId == null) {
                val params = getSessionParams(packageName, size)
                installer.createSession(params)
            } else {
                sessionId
            }
        } catch (e: Exception) {
            log.error(e) { "Error when creating session: " }
            cont.resume(InstallState.Error("${e::class.java.simpleName} ${e.message}"))
            return@suspendCancellableCoroutine
        }
        // set-up receiver for install result
        val receiver = InstallBroadcastReceiver(sessionId) { status, intent, msg ->
            context.unregisterReceiver(this)
            when (status) {
                PackageInstaller.STATUS_SUCCESS -> {
                    val newState = InstallState.Installed(
                        name = state.name,
                        versionName = state.versionName,
                        currentVersionName = state.currentVersionName,
                        lastUpdated = state.lastUpdated,
                        iconDownloadRequest = state.iconDownloadRequest,
                    )
                    cont.resume(newState)
                }
                PackageInstaller.STATUS_PENDING_USER_ACTION -> {
                    val flags = if (SDK_INT >= 31) {
                        FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE
                    } else {
                        FLAG_UPDATE_CURRENT
                    }
                    val pendingIntent =
                        PendingIntent.getActivity(context, sessionId, intent, flags)
                    val progress = installer.getSessionInfo(sessionId)?.progress
                        ?: error("No session info for $sessionId")
                    cont.resume(
                        InstallState.UserConfirmationNeeded(
                            state = state,
                            sessionId = sessionId,
                            intent = pendingIntent,
                            progress = progress,
                        )
                    )
                }
                else -> {
                    if (status == PackageInstaller.STATUS_FAILURE_ABORTED) {
                        cont.resume(InstallState.UserAborted)
                    } else if (status == PackageInstaller.STATUS_FAILURE &&
                        msg != null &&
                        msg.contains("PreapprovalDetails")
                    ) {
                        val newState = InstallState.PreApproved(
                            name = state.name,
                            versionName = state.versionName,
                            currentVersionName = state.currentVersionName,
                            lastUpdated = state.lastUpdated,
                            iconDownloadRequest = state.iconDownloadRequest,
                            result = PreApprovalResult.Error(msg),
                        )
                        cont.resume(newState)
                    } else {
                        cont.resume(InstallState.Error(msg))
                    }
                }
            }
        }
        registerReceiver(
            context,
            receiver,
            IntentFilter(ACTION_INSTALL),
            RECEIVER_NOT_EXPORTED
        )
        cont.invokeOnCancellation {
            log.info { "App installation was cancelled, unregistering broadcast receiver..." }
            context.unregisterReceiver(receiver)
            try {
                installer.abandonSession(sessionId)
            } catch (e: SecurityException) {
                // this can happen if the cancellation came too late and session already concluded
                log.warn(e) { "Error while abandoning session: " }
            }
        }
        // do the actual installation
        try {
            installer.openSession(sessionId).use { session ->
                apkFile.inputStream().use { inputStream ->
                    session.openWrite(packageName, 0, size).use { outputStream ->
                        inputStream.copyTo(outputStream)
                        session.fsync(outputStream)
                    }
                }
                val sender = getInstallIntentSender(sessionId, packageName)
                log.info { "Committing session..." }
                session.commit(sender)
            }
        } catch (e: Exception) {
            log.error(e) { "Error during install session: " }
            cont.resume(InstallState.Error("${e::class.java.simpleName} ${e.message}"))
        }
    }

    suspend fun requestUserConfirmation(
        installState: InstallState.UserConfirmationNeeded,
    ): InstallState = suspendCancellableCoroutine { cont ->
        val receiver = InstallBroadcastReceiver(installState.sessionId) { status, intent, msg ->
            context.unregisterReceiver(this)
            when (status) {
                PackageInstaller.STATUS_SUCCESS -> {
                    val newState = InstallState.Installed(
                        name = installState.name,
                        versionName = installState.versionName,
                        currentVersionName = installState.currentVersionName,
                        lastUpdated = installState.lastUpdated,
                        iconDownloadRequest = installState.iconDownloadRequest,
                    )
                    cont.resume(newState)
                }
                PackageInstaller.STATUS_PENDING_USER_ACTION -> {
                    error("Got STATUS_PENDING_USER_ACTION again")
                }
                else -> {
                    if (status == PackageInstaller.STATUS_FAILURE_ABORTED) {
                        cont.resume(InstallState.UserAborted)
                    } else {
                        cont.resume(InstallState.Error(msg))
                    }
                }
            }
        }
        registerReceiver(
            context,
            receiver,
            IntentFilter(ACTION_INSTALL),
            RECEIVER_NOT_EXPORTED,
        )
        cont.invokeOnCancellation {
            context.unregisterReceiver(receiver)
        }
        installState.intent.send()
    }

    private fun getSessionParams(packageName: String, size: Long? = null): SessionParams {
        val params = SessionParams(SessionParams.MODE_FULL_INSTALL)
        params.setAppPackageName(packageName)
        size?.let { params.setSize(it) }
        params.setInstallLocation(PackageInfo.INSTALL_LOCATION_AUTO)
        if (SDK_INT >= 26) {
            params.setInstallReason(PackageManager.INSTALL_REASON_USER)
        }
        if (SDK_INT >= 31) {
            params.setRequireUserAction(SessionParams.USER_ACTION_NOT_REQUIRED)
        }
        if (SDK_INT >= 33) {
            params.setPackageSource(PackageInstaller.PACKAGE_SOURCE_STORE)
        }
        if (SDK_INT >= 34) {
            // Once the update ownership enforcement is enabled,
            // the other installers will need the user action to update the package
            // even if the installers have been granted the INSTALL_PACKAGES permission.
            // The update ownership enforcement can only be enabled on initial installation.
            // Set this to true on package update is a no-op.
            params.setRequestUpdateOwnership(true)
        }
        return params
    }

    private fun getInstallIntentSender(
        sessionId: Int,
        packageName: String,
    ): IntentSender {
        // Don't use a different action for preapproval and installation,
        // because Android sometimes sends installation broadcasts to preapproval intent.
        val broadcastIntent = Intent(ACTION_INSTALL).apply {
            setPackage(context.packageName)
            putExtra(EXTRA_SESSION_ID, sessionId)
            putExtra(EXTRA_PACKAGE_NAME, packageName)
            addFlags(Intent.FLAG_RECEIVER_FOREGROUND)
        }
        // intent flag needs to be mutable, otherwise the intent has no extras
        val flags = if (SDK_INT >= 31) FLAG_UPDATE_CURRENT or FLAG_MUTABLE else FLAG_UPDATE_CURRENT
        val pendingIntent = PendingIntent.getBroadcast(context, sessionId, broadcastIntent, flags)
        return pendingIntent.intentSender
    }
}
